### DISCUSSION TITLE
vcnt output from grid stat

### DISCUSSION THREAD

**Discussion by fearon10** [dtcenter/METplus#1970] [Configuration] [answered]
Hi There, 

I am struggling to interpret the output in ...vcnt.txt, specifically the number of entries/rows for each column, even after reviewing the section entitled Appendix G Vectors and Vector Statistics. I have also scanned a couple other related discussions on the vcnt topic but did not resolve my issue.

I have successfully run grid stat for U and V wind component variables for one pressure level, e.g., 500 hPa, for one forecast hour. In looking through the "vcnt.txt" output file (attached), I find that I have two entries for each VX_MASK region. For example, the column MSVE, I have values of 1.23 and 1.46 for VX_MASK=FULL or global. These values seem reasonable and represent the "Mean squared length of the vector difference between the forecast and observed winds" I see my FCST_VAR and OBS_VAR are both UGRD_VGRD. Therefore, why do I have two values/entries for each VX_MASK. Shouldn't there be one value? Apologies if I am missing something obvious in the documentation. I thought maybe each entry was a component and they need to be combined? Any help would be much appreciated. 

Thank you,
Matt

[grid_stat_suite_NEPv1_myn_tie_tho_s5m_gfsv16ics_GFS_000000L_20210701_000000V_vcnt.txt](https://github.com/dtcenter/METplus/files/10078167/grid_stat_suite_NEPv1_myn_tie_tho_s5m_gfsv16ics_GFS_000000L_20210701_000000V_vcnt.txt)

**Comment by CPKalb** [comment, marked as answer]
Hi Matt:

When I look at your output file, the TOTAL is different for every masking regions.  For example, the first "FULL" region has a total number of points of 142,953, while the second has a total of 97,479.  This means that the data in the domains are different, although they have the same name.

Would you be able to post your configuration file (either the MET or METplus configuration)?  To understand why there are two regions with the same labels but different numbers of points, I need to see some of the settings in the configuration file.

Thanks,
Christina

**Reply by fearon10** [reply]
Hi Christina,
Thank you for your quick reply. Please find my configuration file attached.  

Thanks again for your assistance.
Matt

[GridStatConfig_suite_NEPv1_myn_tie_tho_s5m_gfsv16ics_2021070100_000000.txt](https://github.com/dtcenter/METplus/files/10079991/GridStatConfig_suite_NEPv1_myn_tie_tho_s5m_gfsv16ics_2021070100_000000.txt)

**Reply by JohnHalleyGotway** [reply]
Hello @fearon10, I work with @CPKalb on providing support for METplus. This sure is a curious result. Looking at your config file, I have a guess as to what's happening. You've set the following inside both the `fcst` and `obs` dictionaries:
```
     is_u_wind     = TRUE;
     is_v_wind     = TRUE;
```
But you put it at the top-level of config file context in those dictionaries rather than inside the `field` array entries. Anything defined at that level of context applies to ALL fields within the dictionary. So basically, all the fields are being identified as containing both the U component of wind AND the V component! That obviously doesn't make sense, but surprisingly, Grid-Stat ran and didn't produce an error! One of the lines contains real data while the other is bogus.

Sure enough, I was able to replicate this behavior by configuring Grid-Stat in the same way using sample test data.

Based on this, I do think we should add another check to make sure that a single field is not identified as being both U and V wind. Hopefully, if you'd received a useful error message, it would have pointed you in the right direction.

The fix is easy. Please just modify your config file as shown below. Note that I factored out the threshold settings to the top-level config file context since you want to apply them to both the forecast and observation datasets.

```
cat_thresh    = [];
censor_thresh = [];
censor_val    = [];
mpr_column    = [];
mpr_thresh    = [];
cnt_thresh    = [ NA ];
cnt_logic     = UNION;
wind_thresh   = [ >= 10 ];
wind_logic    = UNION;

fcst = {
     file_type = NETCDF_NCCF;
     field = [
        {
           name      = "zonal_wind";
           level     = ["(0,4,*,*)"];
           is_u_wind = TRUE;
        },
        {
           name      = "meridional_wind";
           level     = ["(0,4,*,*)"];
           is_v_wind = TRUE;
        }
     ];
}
obs = {
   field = [
      {
         name      = "UGRD";
         level     = ["P500"];
         is_u_wind = TRUE;
      },
      {
         name      = "VGRD";
         level     = ["P500"];
         is_v_wind = TRUE;
      }
   ];
}
```

**Reply by JohnHalleyGotway** [reply]
Here's the MET issue I wrote up to add the sanity check: dtcenter/MET#2357

**Reply by fearon10** [reply]
Hello @JohnHalleyGotway and @CPKalb, 
Thank you both for your assistance. I would not have discovered that the is_u_wind and is_v_wind settings applied to all variables when specified outside of the field strings. However, thinking about it now, that makes sense. Upon receiving your latest comment @JohnHalleyGotway, I tried a new test with is_u_wind and is_v_wnd set within the fcst and obs field strings and got the correct result. 
Thanks again,
Matt

**Reply by JohnHalleyGotway** [reply]
Great, thanks for confirming!

FYI, for dtcenter/MET#2357, I'll add the sanity check logic which would result in this error message being printed:
```
ERROR  : 
ERROR  : VarInfo::set_dict() -> At most one wind attribute flag (is_u_wind, is_v_wind, is_wind_speed, is_wind_direction) can be set to true for each field.
ERROR  :
```
